core/consensus/pow/primitives/src/lib.rs

```diff
@@ -0,0 +1,29 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use substrate_client::decl_runtime_apis;
+use rstd::vec::Vec;
+use sr_primitives::ConsensusEngineId;
+use primitives::H256;
+
+/// The `ConsensusEngineId` of PoW.
+pub const POW_ENGINE_ID: ConsensusEngineId = [b'p', b'o', b'w', b'_'];
+
+/// Type of difficulty.
+pub type Difficulty = u128;
+
+/// Type of seal.
+pub type Seal = Vec<u8>;
+
+decl_runtime_apis! {
+	/// API necessary for block authorship with Proof of Work.
+	pub trait PowApi {
+		/// Verify a given proof of work against the current difficulty.
+		/// Note that `pre_hash` is always a hash of a direct child.
+		///
+		/// Returns the current difficulty.
+		fn verify(pre_hash: &H256, seal: &Seal) -> Option<Difficulty>;
```

Comment:
Can we use this API for implementing something like ethash? If so, then I guess the type of difficulty should be able to fit `H256`, I guess, right?

Reply:
The current difficulty type is `u128`, which I *think* should be more than enough. We should be able to use this to fit nearly all mining algorithms (including ethash). The current difficulty on Ethereum is around 1000x less than `u64::max_value()`. The current total difficulty of Ethereum is more than `u64::max_value()`, but it's way less than `u128::max_value()`, and I think it will never have a chance to exceed it in any reasonable timespan.

Practically using either `u128` or `U256` should be fine. But m concern for changing this to `U256` is that it's less nice compared with `u128`, which is a native type so easier to use.

Reply:
The argument about `u64` w.r.t to ethash is quite convicing for me and I of course deter to your discretion about other PoW algos. 

But it would be great if you added this rationale to the docs!